2.2-taptweak.ipynb

```diff
@@ -170,38 +175,40 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "# Generate a key pair\n",
-    "x, P = generate_key_pair()\n",
-    "print(\"Private key: {}\\nPublic key: {}\\n\".format(x.secret, P.get_bytes().hex()))\n",
+    "# Alice generates a key pair\n",
+    "x_key, P_key = generate_key_pair()\n",
+    "print(\"Private key: {}\\nPublic key: {}\\n\".format(x_key.secret, P_key.get_bytes().hex()))\n",
     "\n",
-    "# Tweak the public key\n",
-    "t = random.randrange(1, SECP256K1_ORDER)\n",
-    "print(\"Tweak: {}\".format(t))\n",
-    "Q = P.tweak_add(t)\n",
+    "# Alice generates the tweak from the contract\n",
+    "contract = \"Alice pays 10 BTC to Bob\"\n",
+    "t = sha256(contract.encode('utf-8'))\n",
+    "print(\"Tweak from original contract: {}\\n\".format(t.hex()))\n",
+    "\n",
+    "# Alice tweaks her key pair\n",
+    "Q_key = P_key.tweak_add(t)\n",
+    "q_key = x_key.add(t)\n",
+    "print(\"Tweaked private key: {}\\nTweaked public key: {}\\n\".format(q_key.secret, Q_key.get_bytes().hex()))\n",
     "\n",
-    "# Create a fake tweak\n",
-    "t2 = random.randrange(1, SECP256K1_ORDER)\n",
-    "print(\"Tweak 2: {}\\n\".format(t2))\n",
+    "# Alice produces a valid signature for this tweaked public key\n",
+    "msg = sha256(b'I agree to the committed contract')\n",
+    "sig = q_key.sign_schnorr(msg)\n",
     "\n",
-    "# Solve: x` = x - t' + t\n",
-    "x_int = x.as_int()\n",
-    "x2_int = (x_int - t2 + t) % SECP256K1_ORDER\n",
+    "# Given P, Q and the contract, Bob believes he can verify the following:\n",
     "\n",
-    "x2_key = x = ECKey().set(x2_int, True)\n",
-    "P2 = x2_key.get_pubkey()\n",
-    "Q2 = P2.tweak_add(t2)\n",
+    "# 1) The contract 'Alice pays 10 BTC to Bob' is committed to public key Q\n",
+    "print(\"'Alice pays 10 BTC to Bob' appears to be committed to Q: {}\".format(\\\n",
```

Comment:
Nit: Could use previously defined `contract` variable

Reply:
Originally kept as string literal, since Bob or verifier must perform his/her own computation, but either way fine for me.

Reply:
Understood. Good to leave as is then.